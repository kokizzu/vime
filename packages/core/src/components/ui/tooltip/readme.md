# vime-tooltip

A small pop-up box that appears when a user moves their mouse over an element. Their main purpose
is to provide a description about the function of that element.

## Visual

<img
  src="https://raw.githubusercontent.com/vime-js/vime/master/packages/core/src/components/ui/tooltip/tooltip.png"
  alt="Vime tooltip component"
/>

<!-- Auto Generated Below -->

## Usage

### Angular

```html {7} title="example.html"
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-controls>
      <vime-control>
        <vime-tooltip>Title</vime-tooltip>
      </vime-control>
    </vime-controls>
  </vime-ui>
</vime-player>
```

### Html

```html {7}
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-controls>
      <vime-control>
        <vime-tooltip>Title</vime-tooltip>
      </vime-control>
    </vime-controls>
  </vime-ui>
</vime-player>
```

### React

```tsx {7,18}
import React from 'react';
import {
  VimePlayer,
  VimeUi,
  VimeControls,
  VimeControl,
  VimeTooltip,
} from '@vime/react';

function Example() {
  return render(
    <VimePlayer>
      {/* ... */}
      <VimeUi>
        {/* ... */}
        <VimeControls>
          <VimeControl>
            <VimeTooltip>Title</VimeTooltip>
          </VimeControl>
        </VimeControls>
      </VimeUi>
    </VimePlayer>
  );
}
```

### Svelte

```html {7,19} title="example.svelte"
<VimePlayer>
  <!-- ... -->
  <VimeUi>
    <!-- ... -->
    <VimeControls>
      <VimeControl>
        <VimeTooltip>Title</VimeTooltip>
      </VimeControl>
    </VimeControls>
  </VimeUi>
</VimePlayer>

<script lang="ts">
  import {
    VimePlayer,
    VimeUi,
    VimeControls,
    VimeControl,
    VimeTooltip,
  } from '@vime/svelte';
</script>
```

### Vue

```html {8,21,30} title="example.vue"
<template>
  <VimePlayer>
    <!-- ... -->
    <VimeUi>
      <!-- ... -->
      <VimeControls>
        <VimeControl>
          <VimeTooltip>Title</VimeTooltip>
        </VimeControl>
      </VimeControls>
    </VimeUi>
  </VimePlayer>
</template>

<script>
  import {
    VimePlayer,
    VimeUi,
    VimeControls,
    VimeControl,
    VimeTooltip,
  } from '@vime/vue';

  export default {
    components: {
      VimePlayer,
      VimeUi,
      VimeControls,
      VimeControl,
      VimeTooltip,
    },
  };
</script>
```

## Properties

| Property    | Attribute   | Description                                                                                                                           | Type                             | Default     |
| ----------- | ----------- | ------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------- | ----------- |
| `active`    | `active`    | Whether the tooltip is visible or not.                                                                                                | `boolean`                        | `false`     |
| `direction` | `direction` | Determines if the tooltip should grow according to its contents to the left/right. By default content grows outwards from the center. | `"left" \| "right" \| undefined` | `undefined` |
| `hidden`    | `hidden`    | Whether the tooltip is displayed or not.                                                                                              | `boolean`                        | `false`     |
| `position`  | `position`  | Determines if the tooltip appears on top/bottom of it's parent.                                                                       | `"bottom" \| "top"`              | `'top'`     |

## Slots

| Slot | Description                                  |
| ---- | -------------------------------------------- |
|      | Used to pass in the contents of the tooltip. |

## CSS Custom Properties

| Name                         | Description                                                      |
| ---------------------------- | ---------------------------------------------------------------- |
| `--tooltip-bg`               | The background color of the tooltip.                             |
| `--tooltip-border-radius`    | The border radius of the tooltip.                                |
| `--tooltip-box-shadow`       | The box shadow cast around the tooltip.                          |
| `--tooltip-color`            | The text color of the tooltip.                                   |
| `--tooltip-fade-duration`    | The length in seconds that the tooltip will take to fade in/out. |
| `--tooltip-fade-timing-func` | The transition timing function for fading in and out.            |
| `--tooltip-font-size`        | The font size of the tooltip text.                               |
| `--tooltip-padding`          | The padding inside the tooltip.                                  |
| `--tooltip-spacing`          | The space between the tooltip and its parent when active.        |

## Dependencies

### Used by

- [vime-caption-control](../controls/caption-control)
- [vime-fullscreen-control](../controls/fullscreen-control)
- [vime-mute-control](../controls/mute-control)
- [vime-pip-control](../controls/pip-control)
- [vime-playback-control](../controls/playback-control)
- [vime-scrubber-control](../controls/scrubber-control)
- [vime-settings-control](../controls/settings-control)

### Graph

```mermaid
graph TD;
  vime-caption-control --> vime-tooltip
  vime-fullscreen-control --> vime-tooltip
  vime-mute-control --> vime-tooltip
  vime-pip-control --> vime-tooltip
  vime-playback-control --> vime-tooltip
  vime-scrubber-control --> vime-tooltip
  vime-settings-control --> vime-tooltip
  style vime-tooltip fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
