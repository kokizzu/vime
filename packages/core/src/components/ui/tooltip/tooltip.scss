vime-tooltip {
  /**
   * @prop --tooltip-bg: The background color of the tooltip.
   * @prop --tooltip-color: The text color of the tooltip.
   * @prop --tooltip-box-shadow: The box shadow cast around the tooltip.
   * @prop --tooltip-border-radius: The border radius of the tooltip.
   * @prop --tooltip-font-size: The font size of the tooltip text.
   * @prop --tooltip-padding: The padding inside the tooltip.
   * @prop --tooltip-spacing: The space between the tooltip and its parent when active.
   * @prop --tooltip-fade-duration: The length in seconds that the tooltip will take to fade in/out.
   * @prop --tooltip-fade-timing-func: The transition timing function for fading in and out.
   */

  left: 50%;
  transform: translateX(-50%);
  line-height: 1.3;
  pointer-events: none;
  position: absolute;
  opacity: 0;
  white-space: nowrap;
  visibility: hidden;
  background: var(--tooltip-bg);
  border-radius: var(--tooltip-border-radius);
  box-shadow: var(--tooltip-box-shadow);
  color: var(--tooltip-color);
  font-size: var(--tooltip-font-size);
  padding: var(--tooltip-padding);
  transition: opacity var(--tooltip-fade-duration) var(--tooltip-fade-timing-func);
  z-index: $z-index-tooltip;

  &[aria-hidden='false'] {
    opacity: 1;
    visibility: visible;
  }

  &.hidden {
    display: none;
  }

  &.onTop {
    bottom: 100%;
    margin-bottom: var(--tooltip-spacing);
  }

  &.onBottom {
    top: 100%;
    margin-top: var(--tooltip-spacing);
  }

  &.growLeft {
    left: auto;
    right: 0;
    transform: none;
  }

  &.growRight {
    left: 0;
    transform: none;
  }
}