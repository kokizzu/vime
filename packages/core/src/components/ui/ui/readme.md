# vime-ui

Simple container that holds a collection of user interface components.

The only important role this component really has is, avoiding overlapping custom UI with the
native iOS media player UI. Therefore, custom UI is only displayed on iOS if the `playsinline` prop
is `true`, and the player is not in fullscreen mode.

<!-- Auto Generated Below -->

## Usage

### Angular

```html {3-5} title="example.html"
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
  </vime-ui>
</vime-player>
```

### Html

```html {3-5}
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
  </vime-ui>
</vime-player>
```

### React

```tsx {2,8}
import React from 'react';
import { VimePlayer, VimeUi } from '@vime/react';

function Example() {
  return render(
    <VimePlayer>
      {/* ... */}
      <VimeUi>{/* ... */}</VimeUi>
    </VimePlayer>
  );
}
```

### Vue

```html {4-6,11,16} title="example.vue"
<template>
  <VimePlayer>
    <!-- ... -->
    <VimeUi>
      <!-- ... -->
    </VimeUi>
  </VimePlayer>
</template>

<script>
  import { VimePlayer, VimeUi } from '@vime/vue';

  export default {
    components: {
      VimePlayer,
      VimeUi,
    },
  };
</script>
```

## Dependencies

### Used by

- [vime-default-ui](../default-ui)

### Graph

```mermaid
graph TD;
  vime-default-ui --> vime-ui
  style vime-ui fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
