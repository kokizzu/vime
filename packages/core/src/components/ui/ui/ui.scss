vime-ui {
  position: relative;
  pointer-events: none;
  z-index: $z-index-ui;
  width: 100%;

  &.hidden {
    display: none;
  }

  &.video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
  }
}