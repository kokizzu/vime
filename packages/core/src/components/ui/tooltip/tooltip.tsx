import {
  h, Host, Component, Element, Prop,
} from '@stencil/core';
import { isString } from '../../../utils/unit';
import { withPlayerContext } from '../../core/player/PlayerContext';
import { PlayerProps } from '../../core/player/PlayerProps';
import { TooltipDirection, TooltipPosition } from './types';

let tooltipIdCount = 0;

/**
 * @slot - Used to pass in the contents of the tooltip.
 */
@Component({
  tag: 'vime-tooltip',
  styleUrl: 'tooltip.scss',
})
export class Tooltip {
  // Avoid tooltips flashing when player initializing.
  private hasLoaded = false;

  @Element() el!: HTMLVimeTooltipElement;

  /**
   * Whether the tooltip is displayed or not.
   */
  @Prop() hidden = false;

  /**
   * Whether the tooltip is visible or not.
   */
  @Prop() active = false;

  /**
   * Determines if the tooltip appears on top/bottom of it's parent.
   */
  @Prop() position: TooltipPosition = 'top';

  /**
   * Determines if the tooltip should grow according to its contents to the left/right. By default
   * content grows outwards from the center.
   */
  @Prop() direction?: TooltipDirection;

  /**
   * @internal
   */
  @Prop() isTouch: PlayerProps['isTouch'] = false;

  componentDidLoad() {
    this.hasLoaded = true;
  }

  private getId() {
    // eslint-disable-next-line prefer-destructuring
    const id = this.el.id;
    if (isString(id) && id.length > 0) return id;
    tooltipIdCount += 1;
    return `vime-tooltip-${tooltipIdCount}`;
  }

  render() {
    return (
      <Host
        id={this.getId()}
        role="tooltip"
        aria-hidden={(!this.active || this.isTouch) ? 'true' : 'false'}
        class={{
          hidden: !this.hasLoaded || this.hidden,
          onTop: (this.position === 'top'),
          onBottom: (this.position === 'bottom'),
          growLeft: (this.direction === 'left'),
          growRight: (this.direction === 'right'),
        }}
      >
        <slot />
      </Host>
    );
  }
}

withPlayerContext(Tooltip, [
  'isTouch',
]);
