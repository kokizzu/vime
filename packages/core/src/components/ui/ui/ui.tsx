import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { withPlayerContext } from '../../core/player/PlayerContext';
import { PlayerProps } from '../../core/player/PlayerProps';
import { IS_IOS } from '../../../utils/support';

/**
 * @slot - Used to pass in UI components for the player.
 */
@Component({
  tag: 'vime-ui',
  styleUrl: 'ui.scss',
})
export class UI {
  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps['isVideoView'] = false;

  /**
   * @internal
   */
  @Prop() playsinline: PlayerProps['playsinline'] = false;

  /**
   * @internal
   */
  @Prop() isFullscreenActive: PlayerProps['isFullscreenActive'] = false;

  render() {
    const canShowCustomUI = !IS_IOS
      || !this.isVideoView
      || (this.playsinline && !this.isFullscreenActive);

    return (
      <Host
        class={{
          hidden: !canShowCustomUI,
          video: this.isVideoView,
        }}
      >
        <div>
          {canShowCustomUI && <slot />}
        </div>
      </Host>
    );
  }
}

withPlayerContext(UI, [
  'isVideoView',
  'playsinline',
  'isFullscreenActive',
]);
