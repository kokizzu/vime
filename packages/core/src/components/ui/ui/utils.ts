import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';

export const findUIRoot = (ref: any) => {
  let ui = getElement(ref);

  while (!(/^VIME-UI$/.test(ui?.nodeName))) {
    ui = ui.parentElement as HTMLStencilElement;
  }

  return ui as HTMLVimeUiElement;
};
