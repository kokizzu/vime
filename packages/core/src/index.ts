import {
  initialState, isWritableProp, PlayerProp, PlayerProps, WritableProps,
} from './components/core/player/PlayerProps';
import { Dispatcher, createDispatcher } from './components/core/player/PlayerDispatcher';
import { withPlayerContext, usePlayerContext } from './components/core/player/PlayerContext';
import { PlayerEvent, PlayerEvents } from './components/core/player/PlayerEvents';
import { PlayerMethods } from './components/core/player/PlayerMethods';
import { findRootPlayer } from './components/core/player/utils';
import { ViewType } from './components/core/player/ViewType';
import { MediaType } from './components/core/player/MediaType';
import { loadSprite } from './utils/network';

export * from './components';

export {
  PlayerProp,
  PlayerProps,
  PlayerEvent,
  PlayerEvents,
  PlayerMethods,
  ViewType,
  MediaType,
  loadSprite,
  initialState,
  isWritableProp,
  WritableProps,
  findRootPlayer,
  withPlayerContext,
  usePlayerContext,
  Dispatcher,
  createDispatcher,
};
