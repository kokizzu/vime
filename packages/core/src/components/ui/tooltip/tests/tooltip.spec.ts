import { SpecPage } from '@stencil/core/testing';
import { newUISpecPage } from '../../ui/tests';
import { Tooltip } from '../tooltip';

let page: SpecPage;
let tooltip: HTMLVimeTooltipElement;

beforeEach(async () => {
  ({ page } = await newUISpecPage(
    [Tooltip],
    '<vime-tooltip>Play</vime-tooltip>',
  ));

  tooltip = page.root!.querySelector('vime-tooltip')!;
});

it('should be structurally sound', () => {
  expect(tooltip).toMatchSnapshot();
});

it('should not render if hidden', async () => {
  tooltip.hidden = true;
  await page.waitForChanges();
  expect(tooltip).toHaveClass('hidden');
});

it('should toggle aria hidden property', async () => {
  expect(tooltip).toEqualAttribute('aria-hidden', 'true');
  tooltip.active = true;
  await page.waitForChanges();
  expect(tooltip).toEqualAttribute('aria-hidden', 'false');
  tooltip.isTouch = true;
  await page.waitForChanges();
  expect(tooltip).toEqualAttribute('aria-hidden', 'true');
});

it('should set positional classes', async () => {
  tooltip.position = 'top';
  await page.waitForChanges();
  expect(tooltip).toHaveClass('onTop');
  expect(tooltip).not.toHaveClass('onBottom');
  tooltip.position = 'bottom';
  await page.waitForChanges();
  expect(tooltip).not.toHaveClass('onTop');
  expect(tooltip).toHaveClass('onBottom');
});

it('should set directional classes', async () => {
  tooltip.direction = 'left';
  await page.waitForChanges();
  expect(tooltip).toHaveClass('growLeft');
  expect(tooltip).not.toHaveClass('growRight');
  tooltip.direction = 'right';
  await page.waitForChanges();
  expect(tooltip).not.toHaveClass('growLeft');
  expect(tooltip).toHaveClass('growRight');
  tooltip.direction = undefined;
  await page.waitForChanges();
  expect(tooltip).not.toHaveClass('growLeft');
  expect(tooltip).not.toHaveClass('growRight');
});
