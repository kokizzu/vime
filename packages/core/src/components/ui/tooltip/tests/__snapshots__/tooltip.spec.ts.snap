// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound 1`] = `
<vime-tooltip aria-hidden="true" class="hidden onTop" id="vime-tooltip-1" role="tooltip">
  <!---->
  Play
</vime-tooltip>
`;
